### PR TITLE
refactor: enable -Wconversion warning for eval.c

### PR DESCRIPTION
**Pull request by dundargoc**
Work on https://github.com/neovim/neovim/issues/567

### CODE REVIEW

**Review by zeertzjq — COMMENTED** [review]
Isn't it better to change the types of parameters of `buf_byteidx_to_charidx()` and `buf_charidx_to_byteidx()`?

**Comment by dundargoc** [comment]
Note to self: change linenr_T to int before doing this PR.

File: src/nvim/eval.c [hunk 1]
```diff
@@ -8985,8 +8985,8 @@ static void delete_var(hashtab_T *ht, hashitem_T *hi)
 static void list_one_var(dictitem_T *v, const char *prefix, int *first)
 {
   char *const s = encode_tv2echo(&v->di_tv, NULL);
-  list_one_var_a(prefix, (const char *)v->di_key, STRLEN(v->di_key),
-                 v->di_tv.v_type, (s == NULL ? "" : s), first);
+  list_one_var_a(prefix, (const char *)v->di_key, (ptrdiff_t)STRLEN(v->di_key),
+                 (int)v->di_tv.v_type, (s == NULL ? "" : s), first);
```
Comment (zeertzjq) [inline src/nvim/eval.c:8989]: Third argument of `list_one_var_a()` can be changed to `VarType` type

**Comment by zeertzjq** [comment]
Oh it was the fourth argument.

File: src/nvim/eval.c [hunk 2]
```diff
@@ -7612,7 +7612,7 @@ char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
       return NULL;
     }
 
-    char *ret = xmalloc(*len + 1);
+    char *ret = xmalloc((size_t)(*len) + 1);
```
Comment (zeertzjq) [inline src/nvim/eval.c:7615]: `*len` may be a negative number here?

**Pull request merged**